### PR TITLE
[oscar][ray backend] Actor creation

### PR DESCRIPTION
**Pull request by fyrestone**
## What do these changes do?
Actor creation on ray backend.

## Related issue number

https://github.com/mars-project/mars/issues/1903

### CODE REVIEW

File: mars/oscar/__init__.py [hunk 1]
```diff
@@ -12,11 +12,66 @@
...
+
+
+def unregister_actor_implementation(actor_cls):
+    try:
+        del _actor_implementation[id(actor_cls)]
+    except KeyError:
+        pass
+
+
+# Cython has a bug about inspecting async function: https://github.com/cython/cython/issues/2273
```
Comment (wjsi) [inline mars/oscar/__init__.py:40]: We do not need to use ``asyncio.iscoroutinefunction``. Checking return value is enough in most scenarios. Thus please reconsider moving this code into ``__init__``.

File: mars/oscar/backends/ray/context.pyx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            raise ActorAlreadyExist(f'Actor {uid} already exist, cannot create')
+
+        actor_handle = ray.remote(actor_cls).options(name=str(uid)).remote(*args, **kwargs)
+        actor.actor_handle = actor_handle
+        actor.uid = uid
+        actor.address = address
+        self.actors[uid] = actor
+        actor_handle._set_uid.remote(uid)
+        actor_handle._set_address.remote(address)
+        actor_handle.__post_create__.remote()
```
Comment (wjsi) [inline mars/oscar/backends/ray/context.pyx:49]: Is this method synchronous? Can this call utilize asyncio?

File: mars/oscar/backends/ray/context.pyx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+
+cdef class _RayActorHandle:
+    cdef object actor_handle
+    cdef object uid
+    cdef object address
+
+
+cdef class RayActorContext(BaseActorContext):
+    cdef public dict actors
```
Comment (wjsi) [inline mars/oscar/backends/ray/context.pyx:31]: I wonder how can this actor registry be visible to all actor processes? If not, ``has_actor`` may not work.

File: mars/oscar/backends/ray/context.pyx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        if uid in self.actors:
+            raise ActorAlreadyExist(f'Actor {uid} already exist, cannot create')
+
+        actor_handle = ray.remote(actor_cls).options(name=str(uid)).remote(*args, **kwargs)
+        actor.actor_handle = actor_handle
+        actor.uid = uid
+        actor.address = address
+        self.actors[uid] = actor
+        actor_handle._set_uid.remote(uid)
+        actor_handle._set_address.remote(address)
```
Comment (wjsi) [inline mars/oscar/backends/ray/context.pyx:48]: Can address here determine the placement group of the actor?

File: mars/oscar/backends/ray/tests/test_ray_actor_context.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        ref1 = await mo.create_actor(DummyActor, 1)
+        with self.assertRaises(TypeError):
+            await ref1.add(1.0)
+        ref2 = await mo.create_actor(DummyActor, 2)
+        with self.assertRaises(TypeError):
+            await ref1.send(ref2, 'add', 1.0)
+        with self.assertRaises(mo.ActorNotExist):
+            await mo.actor_ref('fake_uid').add(1)
+
+    # Fail on ray backend, the DummyActor.create will create an actor in actor process.
```
Comment (wjsi) [inline mars/oscar/backends/ray/tests/test_ray_actor_context.py:169]: Configuring RayActorContext as the global context when starting an actor in Ray will enable creating actor in actors.

File: mars/oscar/core.pyx [hunk 6]
```diff
@@ -13,8 +13,10 @@
 # limitations under the License.
 
 import asyncio
+import cython
```
Comment (wjsi) [inline mars/oscar/core.pyx:16]: ``cimport cython`` ?

File: mars/oscar/__init__.py (hunk 1 above)
Comment (fyrestone) [reply mars/oscar/__init__.py:40]: But ray uses `asyncio.iscoroutinefunction` to determine whether to run async or not.

File: mars/oscar/backends/ray/context.pyx (hunk 2 above)
Comment (fyrestone) [reply mars/oscar/backends/ray/context.pyx:49]: The `remote()` call just send the method name and args to remote. This method does not block waiting for the result.

File: mars/oscar/backends/ray/context.pyx (hunk 3 above)
Comment (fyrestone) [reply mars/oscar/backends/ray/context.pyx:31]: From the discussion, mars holds a context in each node, the `actors` is not shared among them. Am I right?

File: mars/oscar/__init__.py (hunk 1 above)
Comment (wjsi) [reply mars/oscar/__init__.py:40]: I'd rather add a util to enable ``asyncio.iscoroutinefunction`` by adding a util function like:

```python
# add in a pure .py file
def aio_coro_function(func):
    async def dummy():
        pass
    func._is_coroutine = getattr(dummy, '_is_coroutine')
```

File: mars/oscar/backends/ray/context.pyx (hunk 4 above)
Comment (fyrestone) [reply mars/oscar/backends/ray/context.pyx:48]: Yes, we can use placement group here. But I need to figure out the strategy that how to place the group by address.

File: mars/oscar/core.pyx (hunk 6 above)
Comment (fyrestone) [reply mars/oscar/core.pyx:16]: I can use `cimport` instead of `import`.

Thanks. I will fix it.

File: mars/oscar/__init__.py (hunk 1 above)
Comment (fyrestone) [reply mars/oscar/__init__.py:40]: Then we need to use this `iscoroutinefunction` in ray or monkey patch ray.

File: mars/oscar/backends/ray/context.pyx (hunk 2 above)
Comment (wjsi) [reply mars/oscar/backends/ray/context.pyx:49]: ``create_actor`` need to return when ``__post_create__`` is finished. Thus we need to wait asynchronously.

File: mars/oscar/backends/ray/tests/test_ray_actor_context.py (hunk 5 above)
Comment (fyrestone) [reply mars/oscar/backends/ray/tests/test_ray_actor_context.py:169]: The problem is that `DummyActor, 1` and `DummyActor, 2` are in different processes.
- The RayActorContext in current process has actor `DummyActor, 1`
- The RayActorContext in `DummyActor, 1` has actor `DummyActor, 2`.

Then we get `DummyActor, 2` in current process as `ref2`, any call on `ref2` will be failed. Maybe we can use ray named actor to fix this issue.

File: mars/oscar/__init__.py (hunk 1 above)
Comment (wjsi) [reply mars/oscar/__init__.py:40]: A tested diff can be https://github.com/mars-project/mars/compare/refactor/oscar...wjsi:refactor/oscar

File: mars/oscar/backends/ray/context.pyx (hunk 3 above)
Comment (wjsi) [reply mars/oscar/backends/ray/context.pyx:31]: In my understanding, has_actor is needed to work in the scope of a node with processes hosting actors. I cannot see how Actor A created in Actor B can be seen in Actor C when A, B and C are all inside one node.

File: mars/oscar/__init__.py (hunk 1 above)
Comment (wjsi) [reply mars/oscar/__init__.py:40]: > Then we need to use this `iscoroutinefunction` in ray or monkey patch ray.

Sorry deleted a response by mistake. In my implementation above, no changes needed for ``iscoroutinefunction``.

Comment (fyrestone) [reply mars/oscar/__init__.py:40]: A bit hacky. I will try it, thanks.

File: mars/oscar/backends/ray/context.pyx (hunk 2 above)
Comment (fyrestone) [reply mars/oscar/backends/ray/context.pyx:49]: Got it. I will await the `__post_create__`.

File: mars/oscar/backends/ray/context.pyx (hunk 3 above)
Comment (fyrestone) [reply mars/oscar/backends/ray/context.pyx:31]: It sounds like the named actor is the solution.

**Comment by qinxuye** [comment]
Hi, @fyrestone , we added an actor driver class which has a method called `setup_cluster`, it can be used to setup a cluster for Ray. Besides, new backend now just needs to inherit from `BaseActorBackend`.

File: mars/oscar/__init__.py (hunk 1 above)
Comment (fyrestone) [reply mars/oscar/__init__.py:40]: Ray is using the `inspect. iscoroutinefunction` not the `asyncio. iscoroutinefunction`: https://github.com/ray-project/ray/blob/master/python/ray/_raylet.pyx#L421

Comment (qinxuye) [reply mars/oscar/__init__.py:40]: > Ray is using the `inspect.iscoroutinefunction` not the `asyncio.iscoroutinefunction`: https://github.com/ray-project/ray/blob/master/python/ray/_raylet.pyx#L421
> 
> I can create a PR for ray to use `asyncio.iscoroutinefunction` instead.

Now in #1925 , the Actor class has been moved to `api.py` which is not a cython file any more.

Comment (fyrestone) [reply mars/oscar/__init__.py]: Thanks.

**Comment by wjsi** [comment]
LGTM in general. Please rename tests in ray backend from "\*Local*" into "\*Ray*". What's more, copyright info for `mars/oscar/backends/ray/tests/__init__.py` is missing.

**Comment by fyrestone** [comment]
> backend

Fixed, Thanks.

**Review by qinxuye — APPROVED** [review]
LGTM, thank you for your contribution!

**Pull request merged**